tools/clang/lib/Sema/SemaDXR.cpp

```diff
@@ -1254,12 +1254,151 @@ void DiagnoseClosestHitEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
   return;
 }
 
+static bool NodeInputIsCompatible(StringRef& typeName, StringRef& launchName) {
+  return llvm::StringSwitch<bool>(typeName)
+         .Case("DispatchNodeInputRecord", launchName.equals_lower("broadcasting"))
+         .Case("RWDispatchNodeInputRecord", launchName.equals_lower("broadcasting"))
+         .Case("GroupNodeInputRecords", launchName.equals_lower("coalescing"))
+         .Case("RWGroupNodeInputRecords", launchName.equals_lower("coalescing"))
+         .Case("EmptyNodeInput", launchName.equals_lower("coalescing"))
+         .Case("ThreadNodeInputRecord", launchName.equals_lower("thread"))
+         .Case("RWThreadNodeInputRecord", launchName.equals_lower("thread"))
+         .Default(false);
+}
+
 void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
+  SourceLocation computeLoc = SourceLocation();
+  SourceLocation nodeLoc = SourceLocation();
+  SourceLocation nodeLaunchLoc = SourceLocation();
+  StringRef nodeLaunchType;
+  unsigned inputCount = 0;
+
+  // a function may be both compute and work-graph node
+  for (auto *pAttr : FD->specific_attrs<HLSLShaderAttr>()) {
+    DXIL::ShaderKind shaderKind =
+	    ShaderModel::KindFromFullName(pAttr->getStage());
+    if (shaderKind == DXIL::ShaderKind::Node) {
+      nodeLoc = pAttr->getLocation();
+    } else if (shaderKind == DXIL::ShaderKind::Compute) {
+      computeLoc = pAttr->getLocation();
+    }
+  }
+  // if this isn't a work-graph node we can quit now
+  if (!nodeLoc.isValid())
+    return;
+
+  // save NodeLaunch type for use later
+  if (auto NodeLaunch = FD->getAttr<HLSLNodeLaunchAttr>()) {
+    nodeLaunchType = NodeLaunch->getLaunchType();
+    nodeLaunchLoc = NodeLaunch->getLocation();
+  } else {
+    nodeLaunchType = "broadcasting";
+    nodeLaunchLoc = SourceLocation();
+  }
+
+  // If this is both a compute shader and work-graph node, it may only have
+  // broadcasting launch mode
+  if (computeLoc.isValid() && !nodeLaunchType.equals_lower("broadcasting")) {
+    S.Diags.Report(nodeLaunchLoc, diag::err_hlsl_compute_compatibility)
+      << FD->getName() << nodeLaunchType.lower() + " launch type";
+    S.Diags.Report(computeLoc, diag::note_defined_here) << "compute";
+    // ignore other compute incompatibilities (i.e. input/output records)
```

Comment:
It is a matter of opinion - personally I dislike cascades of secondary error messages as they just confuse matters. In this case if the launch type is incompatible with compute then the shader is invalid - so whether it has input/outputs is academic. I also think that if there are input/outputs it is more likely that the programmer accidentally left compute there rather than accidentally adding the input/outputs.
However, I'm happy to change this if you think it would be more consistent with Clang behavior.

Reply:
The advantage of being more compatible with clang behavior is that the upstream compiler will model clang behavior, so being consistent here will give the same user experience in both places.

Additionally, the Clang behavior is the way it is so that IDE tools can present a comprehensive list of diagnostics rather than stopping at the first error. That is useful in situations where resolving the first error isn't actually enough to get the code to be correct. Getting users stuck in a compile->fail->compile->fail cycle is not a great workflow.

Reply:
I have amended the code to allow further errors here and later at line 1380.
There is also a bail-out in the case of errors in hlsl::DiagnoseTranslationUnit() in SemaDiag.cpp - should that be removed too?